src/java/main/org/apache/zookeeper/server/auth/ProviderRegistry.java

```diff
@@ -61,17 +64,17 @@ public static void initialize() {
         }
     }
 
-    public static AuthenticationProvider getProvider(String scheme) {
+    public static ServerAuthenticationProvider getProvider(ZooKeeperServer zks, String scheme) {
         if(!initialized)
-            initialize();
-        return authenticationProviders.get(scheme);
+            initialize(zks);
+        return WrappedAuthenticationProvider.wrap(authenticationProviders.get(scheme));
```

Comment:
I'm not sure I like this wrapping. In this PR, we are only calling `setZooKeeperServer` in the `initialize` method above, so I'm wondering whether `setZooKeeperServer` really needs to be part of the interface and whether we can make `setZooKeeperServer` just a method of the `ServerAuthenticationProvider` implementation, which we can still set during initialization. 

If we do something like this, then we don't need to deprecate `AuthenticationProvider`. What do you think?